### PR TITLE
NATIVE-17: Display errors

### PR DESCRIPTION
**Pull request by NKlug**

### CODE REVIEW

File: src/modules/common/hocs/withError.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// @flow
```
Comment (maxammann) [inline src/modules/common/hocs/withError.js]: I think we can move the error handling code into an `error/` module if it has no dependencies to the app module.

File: src/modules/common/hocs/withError.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+      if (error) {
+        return <Failure error={new Error(error)} />
+      }
+
+      return <Component {...props} />
+    }
+  }
+
+  // $FlowFixMe Please have a look at this flow warning
```
[the remark this answers is not in the document]
Comment (maxammann) [reply src/modules/common/hocs/withError.js]: I'll look into it.

File: src/routes/categories/containers/CategoriesContainer.js [hunk 3]
```diff
@@ -17,15 +17,15 @@ const mapStateToProps = (state: StateType, ownProps) => {
   const targetCity: string = ownProps.navigation.getParam('city')
   const targetPath: string = ownProps.navigation.getParam('path') || `/${targetCity}/${language}`
 
-  const cities = state.cities.json
+  const cities = state.cities
 
-  if (!cities) {
+  if (!cities.json || cities.error) {
```
Comment (maxammann) [inline src/routes/categories/containers/CategoriesContainer.js]: A check for `!cities` usually doesn't happed, but can happen when the state gets rehydrated in a wrong way.

Comment (maxammann) [reply src/routes/categories/containers/CategoriesContainer.js]: Change to `  if (!cities || !cities.json || cities.error) {`

File: src/modules/common/hocs/withError.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxammann) [reply src/modules/common/hocs/withError.js]: Hm I can not reproduce that flow error

File: src/modules/common/components/Failure.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  t: TFunction
+|}
+
+export class Failure extends React.Component<PropsType> {
+  render () {
+    const {t} = this.props
+    return <ViewContainer>
+      <Icon name='frown-o' size={50} />
+      <Text>
+        {t('loadingFailed')}
```
[the remark this answers is not in the document]
Comment (maxammann) [reply src/modules/common/components/Failure.js]: If the error appears in the console we're fine.

[the remark this answers is not in the document]
Comment (NKlug) [reply src/modules/common/components/Failure.js]: I'm logging it to the console now.

File: src/modules/error/components/Failure.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+type PropsType = {|
+  error: Error,
+  t: TFunction
+|}
+
+export class Failure extends React.Component<PropsType> {
+  render () {
+    const {error, t} = this.props
+
+    console.error(error)
```
Comment (klinzo) [inline src/modules/error/components/Failure.js]: Should we log this error here?

Comment (NKlug) [reply src/modules/error/components/Failure.js]: Is there a better place to log it? I think this is just the most straightforward.

Comment (maxammann) [reply src/modules/error/components/Failure.js]: Console logs should be avoided in production mode: https://facebook.github.io/react-native/docs/performance#using-consolelog-statements

Also test would output the errors. Maybe there is a better place for logging. e.g. the place where the exception was thrown.

File: src/modules/error/components/Failure.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  error: Error,
+  t: TFunction
+|}
+
+export class Failure extends React.Component<PropsType> {
+  render () {
+    const { t } = this.props
+
+    return <ViewContainer>
+      <Icon name='frown-o' size={50} />
```
Comment (maxammann) [inline src/modules/error/components/Failure.js]: I think I only used the material icons so far. You can use this icon instead: https://material.io/tools/icons/?icon=sentiment_very_dissatisfied&style=baseline

**Pull request merged**